projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,267 @@
+
+    // setting up
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    BoBRobotics::Robots::RCCarBot bot;
+    BoBRobotics::GPS::Gps gps;
+    gps.connect(path_linux);
+    BoBRobotics::BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+    std::atomic<bool> shouldQuit{false};
+    std::mutex mex;
+
+
+    time_t rawtime;
+    struct tm * timeinfo;
+    char buffer[80];
+
+    time (&rawtime);
+    timeinfo = localtime(&rawtime);
+
+    strftime(buffer,sizeof(buffer),"%d-%m-%Y %H:%M:%S",timeinfo);
+    std::string str(buffer);
+
+    int seconds_to_run = 30; // default value
+    if (argc > 1) {
+        seconds_to_run = stoi(std::string(argv[1]));
+        std::cout << "running for " << seconds_to_run << " seconds" << std::endl;
+
+    }
+
+
+
+    // check to see if we get valid coordinates by checking GPS quality
+    while (numTrials > 0) {
+        try {
+            if (gps.getGPSData().gpsQuality != BoBRobotics::GPS::GPSQuality::INVALID) {
+                std::cout << " we have a valid measurement" << std::endl;
+                break;
+            }
+        } catch(...) {
+            std::cout << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]" << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+    if (numTrials == 0) {
+        std::cout << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program " << std::endl;
+        exit(1);
+    }
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+    auto gps_time =  gps.getGPSData().time;
+    int gps_hour = gps_time.hour;
+    int gps_minute = gps_time.minute;
+    int gps_second = gps_time.second;
+
+    std::string folderName;
+
+    //std::ostringstream usb_f
+    //usb_folder << "../../../../../media/usb/dataset_images/" << str;
+    //folderName = usb_folder.str();
+
+    folderName = str;
+
+
+
+    const char *c  = folderName.c_str();
+    mkdir(c, 0777);
+    std::cout << "directory " << c << " created" << std::endl;
+
+    auto cam = getPanoramicCamera();
+    const auto cameraRes = cam->getOutputSize();
+    std::cout << "camera initialised" << std::endl;
+
+    // unwrapper for camera
+   // auto unwrapper = cam->createUnwrapper(cv::Size(WIDTH,HEIGHT));
+
+
+    cv::Mat imgOrig;
+    //--------------------------------->>>>>>>>>>> START RECORDING <<<<<<<<<<<<----------------------------
+
+
+    // start camera thread
+    cv::Mat tmpImage, unwrappedImage, resized;
+    cv::Mat outputImage(cv::Size(WIDTH,HEIGHT), CV_8UC3);
+    std::thread readCameraThread(&readCameraThreadFunc, cam.get(), std::ref(tmpImage), std::ref(shouldQuit), std::ref(mex));
+    std::this_thread::sleep_for(std::chrono::milliseconds(3000));
+
+    // create and open file to write gps coordinates
+    std::ofstream coordinates;
```

Comment:
Fstreams don't check for errors by default, so add this to get exceptions when something goes wrong:
```c++
coordinates.exceptions(std::ios::failbit | std::ios::badbit);
```